tests/unit/s2n_self_talk_session_resumption_test.c

```diff
@@ -554,6 +554,96 @@ int main(int argc, char **argv)
+     */
+    {
+        DEFER_CLEANUP(struct s2n_blob ticket = { 0 }, s2n_free);
+
+        struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT);
+        struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER);
+        EXPECT_NOT_NULL(client_conn);
+        EXPECT_NOT_NULL(server_conn);
+
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, tls13_client_config));
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, tls12_server_config));
+
+        /* Create nonblocking pipes */
+        struct s2n_test_io_pair io_pair = { 0 };
+        EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+        EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+        /* Negotiate initial handshake to produce TLS1.2 session ticket */
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+        EXPECT_TRUE(ARE_FULL_HANDSHAKES(client_conn, server_conn));
+        EXPECT_TRUE(IS_ISSUING_NEW_SESSION_TICKET(server_conn));
+        EXPECT_EQUAL(server_conn->actual_protocol_version, S2N_TLS12);
+        EXPECT_EQUAL(client_conn->actual_protocol_version, S2N_TLS12);
+
+        /* Store the TLS1.2 session ticket */
+        size_t tls12_session_ticket_len = s2n_connection_get_session_length(client_conn);
+        EXPECT_SUCCESS(s2n_alloc(&ticket, tls12_session_ticket_len));
+        EXPECT_SUCCESS(s2n_connection_get_session(client_conn, ticket.data, ticket.size));
+
+        /* Prepare client and server for a second connection */
+        EXPECT_OK(s2n_wipe_connections(client_conn, server_conn, &io_pair));
+
+        /* Client sets up a resumption connection with the received TLS1.2 session ticket data */
+        EXPECT_SUCCESS(s2n_connection_set_session(client_conn, ticket.data, ticket.size));
+
+        /* Set server config to TLS1.3 cipher preferences */
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, server_config));
+
+        /* Negotiate second connection */
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+        EXPECT_TRUE(ARE_FULL_HANDSHAKES(client_conn, server_conn));
+        EXPECT_EQUAL(server_conn->actual_protocol_version, S2N_TLS13);
+        EXPECT_EQUAL(client_conn->actual_protocol_version, S2N_TLS13);
+
+        /* since TLS 1.3 was negotiated, and the client hasn't called recv yet
+         * there should be no session ticket available. 512 is an arbitrary
+         * number, nothing should be copied because this operation should fail
+         */
+        EXPECT_EQUAL(s2n_connection_get_session(client_conn, ticket.data, 512), 0);
+
+        /* the server automatically sends a session ticket as part of completing
+         * handshake io, so it isn't necessary to call send again
+         */
+        uint8_t data = 1;
+        s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+        s2n_recv(client_conn, &data, 1, &blocked);
```

Comment:
I think you can use EXPECT_BLOCKED_ON_IO here?

Reply:
Had to move the functions from the early_data_test, so added those to `test.h`.